File: swift/clientlib.swift
import Foundation

class GreifarmClient {
    let serverIP: String
    let port: Int
    
    init(serverIP: String, port: Int = 8005) {
        self.serverIP = serverIP
        self.port = port
    }
    
    // Prüft die Serververbindung
    func checkServerConnection(completion: @escaping (Bool) -> Void) {
        guard let url = URL(string: "http://\(serverIP):\(port)/") else {
            print("Ungültige URL")
            completion(false)
            return
        }
        
        let task = URLSession.shared.dataTask(with: url) { _, response, error in
            if let error = error {
                DispatchQueue.main.async {
                    print("Verbindungsfehler: \(error.localizedDescription)")
                }
                completion(false)
            } else if let httpResponse = response as? HTTPURLResponse {
                DispatchQueue.main.async {
                    if httpResponse.statusCode == 200 {
                        print("Verbindung zum Server erfolgreich (Statuscode: \(httpResponse.statusCode)).")
                    } else {
                        print("Fehlerhafte Antwort vom Server (Statuscode: \(httpResponse.statusCode)).")
                    }
                }
                completion(httpResponse.statusCode == 200)
            }
        }
        task.resume()
    }
    
    func sendServoPositions(x: Int = 90, y: Int = 90, z: Int = 90, completion: @escaping (String) -> Void) {
        checkServerConnection { isConnected in
            if !isConnected {
                completion("Verbindung zum Server konnte nicht hergestellt werden.")
                return
            }
            
            let clampedX = max(90, min(270, x))
            let clampedY = max(90, min(270, y))
            let clampedZ = max(90, min(270, z))
            let data: [String: Any] = ["x": clampedX, "y": clampedY, "z": clampedZ]
            self.sendRequest(data: data, completion: completion)
        }
    }
    
    func toggleMagnet(state: Bool, completion: @escaping (String) -> Void) {
        checkServerConnection { isConnected in
            if !isConnected {
                completion("Verbindung zum Server konnte nicht hergestellt werden.")
                return
            }
            
            let data: [String: Any] = ["magnet": state]
            self.sendRequest(data: data, completion: completion)
        }
    }
    
    func shutdownServer(completion: @escaping (String) -> Void) {
        checkServerConnection { isConnected in
            if !isConnected {
                completion("Verbindung zum Server konnte nicht hergestellt werden.")
                return
            }
            
            let data: [String: Any] = ["shutdown": true]
            self.sendRequest(data: data, completion: completion)
        }
    }
    
    private func sendRequest(data: [String: Any], completion: @escaping (String) -> Void) {
        guard let url = URL(string: "http://\(serverIP):\(port)/") else {
            DispatchQueue.main.async {
                print("Ungültige URL")
                completion("Ungültige URL")
            }
            return
        }
        
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.addValue("application/json", forHTTPHeaderField: "Content-Type")
        
        do {
            request.httpBody = try JSONSerialization.data(withJSONObject: data, options: [])
        } catch {
            DispatchQueue.main.async {
                print("JSON-Fehler: \(error.localizedDescription)")
                completion("JSON-Fehler: \(error.localizedDescription)")
            }
            return
        }
        
        let task = URLSession.shared.dataTask(with: request) { data, response, error in
            if let error = error {
                DispatchQueue.main.async {
                    print("Netzwerkfehler: \(error.localizedDescription)")
                    completion("Netzwerkfehler: \(error.localizedDescription)")
                }
                return
            }
            
            if let httpResponse = response as? HTTPURLResponse {
                DispatchQueue.main.async {
                    print("HTTP-Statuscode: \(httpResponse.statusCode)")
                    if httpResponse.statusCode != 200 {
                        print("Fehlerhafte Antwort vom Server: Statuscode \(httpResponse.statusCode).")
                    }
                }
            }
            
            if let data = data, let responseString = String(data: data, encoding: .utf8) {
                DispatchQueue.main.async {
                    print("Antwort: \(responseString)")
                    completion(responseString)
                }
            } else {
                DispatchQueue.main.async {
                    print("Unerwartete Antwort")
                    completion("Unerwartete Antwort")
                }
            }
        }
        task.resume()
    }
}

// Beispielnutzung
let client = GreifarmClient(serverIP: "192.168.178.151") // IP anpassen
client.sendServoPositions(x: 120, y: 150, z: 200) { response in
    print("Antwort: \(response)")
}
client.toggleMagnet(state: true) { response in
    print("Magnet: \(response)")
}
client.shutdownServer { response in
    print("Shutdown: \(response)")
}

File: lib/clientlib.py
import json
import socket

class GreifarmClient:
    def __init__(self, server_ip, port=8005):
        self.server_ip = server_ip
        self.port = port

    def send_servo_positions(self, x=90, y=90, z=90):
        x = max(90, min(270, x))
        y = max(90, min(270, y))
        z = max(90, min(270, z))
        data = {"x": x, "y": y, "z": z}
        return self._send_request(data)

    def toggle_magnet(self, state):
        data = {"magnet": state}
        return self._send_request(data)

    def shutdown_server(self):
        data = {"shutdown": True}
        return self._send_request(data)

    def _send_request(self, data):
        try:
            with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as s:
                s.connect((self.server_ip, self.port))
                request = "POST / HTTP/1.1\r\n"
                request += "Content-Type: application/json\r\n"
                request += f"Content-Length: {len(json.dumps(data))}\r\n"
                request += "Connection: close\r\n\r\n"
                request += json.dumps(data)
                s.sendall(request.encode('utf-8'))
                response = s.recv(1024)
                return response.decode('utf-8')
        except Exception as e:
            return {"error": str(e)}

# Beispielnutzung
if __name__ == "__main__":
    client = GreifarmClient("192.168.1.100")  # Server-IP anpassen
    print(client.send_servo_positions(x=120, y=150, z=200))
    print(client.toggle_magnet(True))
    print(client.shutdown_server())
